eth_dummy_phy/examples/simple/main/emac2emac.c

```diff
@@ -0,0 +1,297 @@
+        ESP_LOGI(TAG, "Ethernet Stopped");
+        break;
+    default:
+        break;
+    }
+}
+
+/** Event handler for IP_EVENT_ETH_GOT_IP */
+static void got_ip_event_handler(void *arg, esp_event_base_t event_base,
+                                 int32_t event_id, void *event_data)
+{
+    ip_event_got_ip_t *event = (ip_event_got_ip_t *) event_data;
+    const esp_netif_ip_info_t *ip_info = &event->ip_info;
+
+    ESP_LOGI(TAG, "Ethernet Got IP Address");
+    ESP_LOGI(TAG, "~~~~~~~~~~~");
+    ESP_LOGI(TAG, "ETHIP:" IPSTR, IP2STR(&ip_info->ip));
+    ESP_LOGI(TAG, "ETHMASK:" IPSTR, IP2STR(&ip_info->netmask));
+    ESP_LOGI(TAG, "ETHGW:" IPSTR, IP2STR(&ip_info->gw));
+    ESP_LOGI(TAG, "~~~~~~~~~~~");
+
+#if !CONFIG_EXAMPLE_DHCP_SERVER_EN
+    ping_start(ip_info);
+#endif
+}
+
+#if CONFIG_EXAMPLE_RMII_CLK_SOURCE_DEV
+IRAM_ATTR static void gpio_isr_handler(void *arg)
+{
+    BaseType_t high_task_wakeup = pdFALSE;
+    TaskHandle_t task_handle = (TaskHandle_t)arg;
+
+    vTaskNotifyGiveFromISR(task_handle, &high_task_wakeup);
+    if (high_task_wakeup != pdFALSE) {
+        portYIELD_FROM_ISR();
+    }
+}
+#endif // CONFIG_EXAMPLE_RMII_CLK_SOURCE_DEV
+
+void app_main(void)
+{
+#if CONFIG_EXAMPLE_RMII_CLK_SOURCE_DEV
+    // ESP32 device which is RMII CLK source needs to wait with its Ethernet initialization for the "RMII CLK
+    // Sink Device" since the RMII CLK input pin (GPIO0) is also used as a boot strap pin. If the "RMII CLK Source Device" didn't wait,
+    // the "RMII CLK Sink Device" could boot into incorrect mode.
+    esp_rom_gpio_pad_select_gpio(EMAC_CLK_OUT_180_GPIO);
+    gpio_set_pull_mode(EMAC_CLK_OUT_180_GPIO, GPIO_FLOATING); // to not affect GPIO0 (so the Sink Device could be flashed)
+    gpio_install_isr_service(0);
+    esp_rom_gpio_pad_select_gpio(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO);
+    gpio_set_direction(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO, GPIO_MODE_INPUT);
+    gpio_set_pull_mode(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO, GPIO_PULLDOWN_ONLY);
+    gpio_set_intr_type(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO, GPIO_INTR_ANYEDGE);
+    gpio_intr_enable(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO);
+    TaskHandle_t task_handle = xTaskGetHandle(pcTaskGetName(NULL));
+    gpio_isr_handler_add(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO, gpio_isr_handler, task_handle);
+    ESP_LOGW(TAG, "waiting for RMII CLK sink device interrupt");
+    ESP_LOGW(TAG, "if RMII CLK sink device is already running, reset it by `EN` button");
+    while (1) {
+        ulTaskNotifyTake(pdTRUE, portMAX_DELAY);
+        if (gpio_get_level(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO) == 1) {
+            break;
+        }
+    }
+    ESP_LOGI(TAG, "starting Ethernet initialization");
+#else
+    esp_rom_gpio_pad_select_gpio(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO);
+    gpio_set_direction(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO, GPIO_MODE_OUTPUT);
+    gpio_set_pull_mode(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO, GPIO_FLOATING);
+    gpio_set_level(CONFIG_EXAMPLE_CLK_SINK_READY_GPIO, 0);
```

Comment:
maybe we need to add some comments for this, looks like it also serve as kinda "synchronization" after the RMII sink device resets

Reply:
I would move the comment which is currently present under #if CONFIG_EXAMPLE_RMII_CLK_SOURCE_DEV above that statement because it's already describes "sink" and "source" concept.

Reply:
There is also [this](https://github.com/espressif/esp-eth-drivers/pull/21/files/4281fb21a59e58ac5980a5862b787d666538db22#diff-24ef89703d622d4375865e2309e27da4d5d9feccc1220ddd98409cb274eeb699R285) comment to explain Sink reset.